File: python/code/13-dtft.py
import numpy as np
import matplotlib.pyplot as plt

# Given sequence
x = np.array([1, 2, 3, 4, 5])

# Sample indices
n = np.arange(len(x))

# 501 frequencies from 0 to pi
w = np.linspace(0, np.pi, 501)

# DTFT
X = np.zeros(len(w), dtype=complex)

for i in range(len(w)):
    X[i] = np.sum(
        x * np.exp(-1j * w[i] * n)
    )


# Magnitude
magnitude = np.abs(X)

# Phase
phase = np.angle(X)

# Real part
real_part = np.real(X)

# Imaginary part
imaginary_part = np.imag(X)

plt.figure(figsize=(12, 12))

plt.subplot(4, 1, 1)
plt.plot(w, magnitude)
plt.title("Magnitude of DTFT")
plt.xlabel("ω")
plt.ylabel("|X(e^jω)|")
plt.grid(True)

plt.subplot(4, 1, 2)
plt.plot(w, phase)
plt.title("Phase / Angle of DTFT")
plt.xlabel("ω")
plt.ylabel("Angle")
plt.grid(True)

plt.subplot(4, 1, 3)
plt.plot(w, real_part)
plt.title("Real Part of DTFT")
plt.xlabel("ω")
plt.ylabel("Real")
plt.grid(True)

plt.subplot(4, 1, 4)
plt.plot(w, imaginary_part)
plt.title("Imaginary Part of DTFT")
plt.xlabel("ω")
plt.ylabel("Imaginary")
plt.grid(True)

plt.tight_layout()
plt.show()

File: matlab/code/13-dtft.m
clc;
clear;
close all;

% Given sequence
x = [1 2 3 4 5];

% Time indices
n = 0:4;

% 501 equispaced frequencies from 0 to pi
w = linspace(0, pi, 501);

% Initialize DTFT
X = zeros(size(w));

% Calculate DTFT
for k = 1:length(w)

    X(k) = sum(x .* exp(-1i*w(k)*n));

end

% Magnitude
magnitude = abs(X);

% Phase angle
phase = angle(X);

% Real part
real_part = real(X);

% Imaginary part
imaginary_part = imag(X);

% -----------------------------
% Magnitude
% -----------------------------

figure;
plot(w, magnitude);
grid on;

xlabel('\omega (rad/sample)');
ylabel('|X(e^{j\omega})|');
title('Magnitude of DTFT');

% -----------------------------
% Phase
% -----------------------------

figure;
plot(w, phase);
grid on;

xlabel('\omega (rad/sample)');
ylabel('Phase (radians)');
title('Angle of DTFT');

% -----------------------------
% Real Part
% -----------------------------

figure;
plot(w, real_part);
grid on;

xlabel('\omega (rad/sample)');
ylabel('Real Part');
title('Real Part of DTFT');

% -----------------------------
% Imaginary Part
% -----------------------------

figure;
plot(w, imaginary_part);
grid on;

xlabel('\omega (rad/sample)');
ylabel('Imaginary Part');
title('Imaginary Part of DTFT');